src/librustc_typeck/astconv.rs

```diff
@@ -2157,6 +2157,11 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx> + 'o {
 
     /// Returns the `DefId` of the constant parameter that the provided expression is a path to.
     pub fn const_param_def_id(&self, expr: &hir::Expr) -> Option<DefId> {
+        let expr = match &expr.node {
```

Comment:
`ast_const_to_const` already has this sort of logic — you can extract it from there: https://github.com/rust-lang/rust/blob/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/librustc_typeck/astconv.rs#L2189-L2197
(You want to check that there are no leading statements in the block either.)

Reply:
Oops, I saw that when I was making the change but never thought to just extract it. However, this made me realise that `[x; {let _ = 3; N}]` will still ICE.

If a block has statements, `const_param_def_id` will return `None` and that will make `ExprKind::Repeat` have the type `ty:Err`.

Then, when this line is hit:

https://github.com/rust-lang/rust/blob/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/librustc_typeck/check/regionck.rs#L545

It eventually gets to this call to `expr_ty`, which will return an `Err(())` (due to the `ty::Err`) and that will be propagated to the caller.

https://github.com/rust-lang/rust/blob/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/librustc/middle/mem_categorization.rs#L646

Eventually, that results in this bug:

https://github.com/rust-lang/rust/blob/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/librustc_typeck/check/regionck.rs#L554

And then that ends up causing the ICE later on. As far as I can gather, the only way to stop the ICE is to make it so that a `ty::Err` being returned from `expr_ty` isn't fatal - but I'm not sure how best to do that, any thoughts?

Reply:
Actually, thinking about it further, I don't see that we do need to prevent statements preceding the parameter. For instance, this is already allowed: `let x: [(); { let y = 3; y }] = [(), (), ()];`. However, at the moment I think the two places `const_param_def_id` is used assume that if a const parameter exists, it's alone in an expression. Ideally, I think the fix here is to deal with arbitrary expressions involving const parameters (in both instances). This is a more involved fix for the issue, though I think it would also address issues like https://github.com/rust-lang/rust/issues/61368.

Reply:
I think, as in #61368, proper handling of this might be blocked on lazy normalisation. However, we could be consistent with current parameter handling and special-case `{ P }` (but continue to ICE on anything else for now).

Reply:
Alright, that makes sense, I've pushed a commit that extracts the statement logic from `ast_const_to_const`.